File: trip-server/src/services/queryRewriter.ts
import { queryRewriteLog as log } from '../utils/logger'

/**
 * 本地查询改写：从自然语言 query 中提取检索关键词
 *
 * 替代原来的 LLM 改写（每次 ~800ms），节省 ~750ms/次
 *
 * 策略：
 * 1. 去掉常见停用词
 * 2. 提取 2-5 字中文片段
 * 3. 意图映射 → 分类关键词（"吃"→"美食 餐厅"）
 * 4. 去重后返回最多 6 个关键词
 */

const STOP_WORDS = new Set([
  '什么', '怎么', '哪里', '哪个', '哪些', '谁', '何时',
  '推荐', '请问', '我想', '需要', '可以', '有没有', '是不是',
  '帮我', '介绍', '告诉', '知道', '觉得', '想去', '打算',
  '一下', '一下', '吧', '啊', '呢', '吗', '了', '的', '啦',
  '在', '有', '和', '与', '或', '是', '的', '这', '那',
  '去', '来', '到', '从', '给', '为', '对',
  '好玩', '好吃', '好看', '好喝', '有趣', '漂亮', '不错',
  '大概', '多少', '便宜', '方便', '适合',
])

const INTENT_MAP: Array<[RegExp, string]> = [
  [/吃|美食|餐厅|饭馆|菜|辣|甜|口味/i, '美食 餐厅'],
  [/住|住宿|酒店|宾馆|民宿|客栈|房间/i, '住宿 酒店'],
  [/玩|逛|游|景点|景区|公园/i, '景点 游览'],
  [/买|购物|商业街|商场|街/i, '购物 商业街'],
  [/票|门票|收费|价格|价钱|多少钱/i, '门票 价格'],
  [/交通|公交|地铁|打车|开车|停车/i, '交通'],
  [/拍照|摄影|照片|打卡/i, '拍照 打卡'],
  [/历史|文化|博物馆|古迹|文物/i, '文化 历史'],
  [/自然|山|水|湖|海|森林|风景|风/i, '自然 风景'],
  [/夜|晚上|夜景|夜市|酒吧/i, '夜景 夜市'],
  [/亲|子|小孩|儿童|亲子|带孩/i, '亲子 儿童'],
  [/情侣|约会|浪漫/i, '情侣 浪漫'],
  [/便宜|实惠|平价|经济|省钱|免费/i, '平价 实惠'],
  [/高。端|豪华|奢侈|五星|品质/i, '高端 品质'],
  [/室。内|下雨|天气|夏天|避暑|冬天|暖/i, '室内 避暑'],
]

function extractKeywords(query: string, city: string): string[] {
  const result = new Set<string>()

  // 1. 保留城市名
  if (city) result.add(city)

  // 2. 意图映射
  for (const [pattern, kw] of INTENT_MAP) {
    if (pattern.test(query)) {
      for (const word of kw.split(' ')) result.add(word)
    }
  }

  // 3. 保留原始 query 中的景点实体（去掉停用词后的原文）
  const cleanQuery = query.split(new RegExp(`[${' '.concat('的了吧吗呢啊啦在和有与或是这那来到从给为对').split('').join('')}]`, 'g')).filter(Boolean)
  for (const w of cleanQuery) {
    if (w.length >= 2 && /[一-龥]/.test(w) && !STOP_WORDS.has(w)) result.add(w)
  }

  // 4. 通配类别词
  const specialKeys = ['景点', '美食', '酒店', '住宿', '餐厅', '公园', '博物馆', '夜景', '亲子', '平价', '门票', '交通', '购物']
  for (const k of specialKeys) {
    if (query.includes(k)) result.add(k)
  }

  return [...result].slice(0, 6)
}

/**
 * 本地查询改写（替代原来的 LLM 调用）
 *
 * 速度 ~50ms，不再调 DeepSeek API
 */
export async function rewriteQuery(query: string, city?: string): Promise<string> {
  const keywords = extractKeywords(query, city || '')
  if (keywords.length === 0) return query
  const rewritten = keywords.join(' ')
  if (rewritten !== query) {
    log.debug({ original: query, rewritten }, 'query rewritten')
  }
  return rewritten
}

File: trip-server-py/src/services/rag/query_rewriter.py
"""查询改写模块.

提供本地查询改写能力，不依赖 LLM 调用：
- 提取关键词
- 移除停用词
- 可选：调用 LLM 进行查询扩展（如果对话上下文可用）
"""

import logging
import re
from typing import List, Optional

logger = logging.getLogger(__name__)

# 中文停用词表（简化版）
_STOP_WORDS_ZH = {
    "的", "了", "在", "是", "我", "有", "和", "就", "不", "人",
    "都", "一", "一个", "上", "也", "很", "到", "说", "要", "去",
    "你", "会", "着", "没有", "看", "好", "自己", "这", "但", "从",
    "可以", "这个", "当", "本", "如", "就", "让", "把", "还", "用",
    "没", "能", "过", "她", "他", "它", "们", "那", "些", "什么",
    "怎么", "如何", "为什么", "哪", "吗", "呢", "吧", "啊", "嗯",
}

# 英文停用词表（简化版）
_STOP_WORDS_EN = {
    "the", "a", "an", "and", "or", "but", "in", "on", "at", "to",
    "for", "of", "with", "by", "from", "up", "about", "into", "over",
    "after", "beneath", "between", "under", "i", "me", "my", "we",
    "you", "he", "she", "it", "they", "is", "are", "was", "were",
    "be", "been", "being", "have", "has", "had", "do", "does", "did",
    "will", "would", "shall", "should", "can", "could", "may", "might",
}

_STOP_WORDS = _STOP_WORDS_ZH.union(_STOP_WORDS_EN)


def extract_keywords(text: str, max_keywords: int = 10) -> List[str]:
    """从文本中提取关键词（基于分词和停用词过滤）.

    Args:
        text: 输入文本.
        max_keywords: 最大关键词数量.

    Returns:
        List[str]: 关键词列表，按重要性排序。

    Example:
        >>> extract_keywords("我想去北京故宫玩")
        ['北京', '故宫']
    """
    # 简单分词：按空格、标点分割
    words = re.findall(r"[\w]+", text.lower())
    # 过滤停用词和短词
    keywords = [
        w for w in words
        if w not in _STOP_WORDS and len(w) >= 2
    ]
    # 去重并保持顺序
    seen = set()
    unique_keywords = []
    for kw in keywords:
        if kw not in seen:
            seen.add(kw)
            unique_keywords.append(kw)
    return unique_keywords[:max_keywords]


def rewrite_query(
    query: str,
    city: Optional[str] = None,
    context: Optional[str] = None,
) -> str:
    """本地查询改写（不调用 LLM）.

    对查询进行标准化处理：
    1. 移除多余空格和特殊字符
    2. 提取关键词
    3. 如果有城市信息，追加城市前缀

    Args:
        query: 原始查询文本.
        city: 目标城市（可选）.
        context: 对话上下文（可选，预留用于未来 LLM 扩展）.

    Returns:
        str: 改写后的查询文本.

    Example:
        >>> rewrite_query("我想去故宫玩", city="北京")
        '北京 故宫'
    """
    # 1. 清洗文本：保留中文、英文、数字、空格
    cleaned = re.sub(r"[^\w\s]", " ", query)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()

    # 2. 提取关键词
    keywords = extract_keywords(cleaned)

    # 3. 组装改写后的查询
    parts = []
    if city and city not in keywords:
        parts.append(city)
    parts.extend(keywords)

    rewritten = " ".join(parts) if parts else cleaned

    logger.debug("查询改写: '%s' -> '%s'", query, rewritten)
    return rewritten


async def rewrite_query_with_llm(
    query: str,
    city: Optional[str] = None,
    context: Optional[str] = None,
) -> str:
    """使用 LLM 进行查询改写（可选功能）.

    当需要更复杂的查询扩展时使用，例如：
    - 同义词扩展
    - 意图理解
    - 上下文感知改写

    Args:
        query: 原始查询文本.
        city: 目标城市.
        context: 对话上下文.

    Returns:
        str: LLM 改写后的查询文本.

    Note:
        此函数需要依赖 LLM 服务，如果 LLM 不可用则降级到本地改写。
    """
    # TODO: 实现 LLM 查询改写
    # 暂时降级到本地改写
    logger.debug("LLM 查询改写未实现，降级到本地改写")
    return rewrite_query(query, city, context)


def detect_city(query: str) -> Optional[str]:
    """从查询中检测城市名称.

    Args:
        query: 查询文本.

    Returns:
        Optional[str]: 检测到的城市名，如果未检测到返回 None.

    Example:
        >>> detect_city("我想去北京玩")
        '北京'
    """
    # 只匹配已知城市名（避免误匹配）
    city_patterns = [
        r"(北京|上海|广州|深圳|成都|杭州|西安|南京|重庆|天津|武汉|苏州|厦门|昆明|三亚)",
    ]
    for pattern in city_patterns:
        match = re.search(pattern, query)
        if match:
            return match.group(1)
    return None


def detect_intent(query: str) -> str:
    """检测查询意图.

    Args:
        query: 查询文本.

    Returns:
        str: 意图类型，可选值：'scenic'（景点）、'food'（美食）、
             'hotel'（住宿）、'transport'（交通）、'general'（通用）。

    Example:
        >>> detect_intent("北京有什么好吃的")
        'food'
    """
    intent_keywords = {
        "scenic": ["景点", "景区", "好玩", "玩", "旅游", "游记", "风景"],
        "food": ["吃", "美食", "餐厅", "小吃", "特产", "好吃的"],
        "hotel": ["住", "酒店", "住宿", "民宿"],
        "transport": ["交通", "怎么去", "路线", "打车", "地铁"],
    }
    for intent, keywords in intent_keywords.items():
        if any(kw in query for kw in keywords):
            return intent
    return "general"
